toolkit/tools/scheduler/scheduler.go

```diff
@@ -323,57 +326,76 @@ func buildAllNodes(stopOnFailure, canUseCache bool, packagesToRebuild, testsToRe
+	if err != nil {
+		err = fmt.Errorf("error building implicit node resolution priority set:\n%w", err)
+		return
+	}
+	buildState := schedulerutils.NewGraphBuildState(reservedFiles, priorityBuildSet, maxCascadingRebuilds)
 	buildRunsTests := len(pkgGraph.AllTestNodes()) > 0
 	nodesToBuild := schedulerutils.LeafNodes(pkgGraph, graphMutex, goalNode, buildState, useCachedImplicit)
+	// As the build progresses, we will need to keep track of low priority nodes that were set aside.
+	lowPriorityNodes := make([]*pkggraph.PkgNode, 0)
+	fullGraph := pkgGraph
 
 	for {
 		logger.Log.Debugf("Found %d unblocked nodes: %v.", len(nodesToBuild), nodesToBuild)
 
 		// Each node that is ready to build must be converted into a build request and submitted to the worker pool.
-		newRequests := schedulerutils.ConvertNodesToRequests(pkgGraph, graphMutex, nodesToBuild, packagesToRebuild, testsToRerun, buildState, canUseCache)
+		newRequests := schedulerutils.ConvertNodesToRequests(pkgGraph, graphMutex, nodesToBuild, packagesToRebuild, testsToRerun, buildState, canUseCache, allowLowPriorityNodes)
 		for _, req := range newRequests {
 			buildState.RecordBuildRequest(req)
-			// Decide which priority the build should be. Generally we want to get any remote or prebuilt nodes out of the
-			// way as quickly as possible since they may help us optimize the graph early.
-			// Meta nodes may also be blocking something we want to examine and give higher priority (priority inheritance from
-			// the hypothetical high priority node hidden further into the tree)
-			switch req.Node.Type {
-			case pkggraph.TypePreBuilt:
-				channels.PriorityRequests <- req
-
-				// For now all build nodes are of equal priority
-			case pkggraph.TypeGoal:
-				fallthrough
-			case pkggraph.TypePureMeta:
-				fallthrough
-			case pkggraph.TypeLocalRun:
-				fallthrough
-			case pkggraph.TypeRemoteRun:
-				fallthrough
-			case pkggraph.TypeLocalBuild:
+			// Queue nodes based on how likely they are to help optimize the build.
+			// - High priority nodes are expected to resolve implicit provides that are needed to build other packages.
+			// - Medium priority nodes are our desired build goal.
+			// - Low priority nodes are packages that are not needed to build the desired goal, but may be needed to
+			//    resolve implicit provides if the scheduler is unable to find a provider for them.
+			switch buildState.GetNodePriority(req.Node) {
+			case schedulerutils.HighNodePriority:
+				logger.Log.Tracef("Priority priority request (%d) for node '%s'", buildState.GetNodePriority(req.Node), req.Node.FriendlyName())
+				channels.HighPriorityRequests <- req
+			case schedulerutils.MediumNodePriority:
+				logger.Log.Tracef("Medium priority request (%d) for node '%s'", buildState.GetNodePriority(req.Node), req.Node.FriendlyName())
+				channels.MediumPriorityRequests <- req
+			case schedulerutils.LowNodePriority:
 				fallthrough
 			default:
```

Comment:
what is the default case here, haven't all requests been assigned a priority at this point?

Reply:
Fair point, probably better to throw an error here instead. Will fix.